opentelemetry-sdk/src/opentelemetry/sdk/metrics/__init__.py

```diff
@@ -0,0 +1,307 @@
+        label_keys: Tuple[str, ...] = None,
+        enabled: bool = False,
+        monotonic: bool = False,
+    ):
+        super(Measure, self).__init__(
+            name,
+            description,
+            unit,
+            value_type,
+            label_keys=label_keys,
+            enabled=enabled,
+            monotonic=monotonic,
+        )
+
+    def get_handle(
+        self, label_values: Tuple[str, ...]
+    ) -> metrics_api.MeasureHandle:
+        """See `opentelemetry.metrics.Measure.get_handle`."""
+        handle = self.handles.get(
+            label_values,
+            MeasureHandle(self.value_type, self.enabled, self.monotonic),
+        )
+        self.handles[label_values] = handle
+        return handle
+
+
+class CounterHandle(metrics_api.CounterHandle):
+    def __init__(
+        self, value_type: metrics_api.ValueType, enabled: bool, monotonic: bool
+    ):
+        self.data = 0
+        self.value_type = value_type
+        self.enabled = enabled
+        self.monotonic = monotonic
+
+    def update(self, value: metrics_api.ValueType) -> None:
+        if not self.enabled:
+            return
+        if not self.monotonic and value < 0:
+            logger.warning("Monotonic counter cannot descend.")
+            return
+        if not isinstance(value, self.value_type):
+            logger.warning(
+                "Invalid value passed for %s.", self.value_type.__name__
+            )
+            return
+        self.data += value
+
+    def add(self, value: metrics_api.ValueType) -> None:
+        """See `opentelemetry.metrics.CounterHandle._add`."""
+        self.update(value)
+
+
+class GaugeHandle(metrics_api.GaugeHandle):
+    def __init__(
+        self,
+        value_type: Union[Type[float], Type[int]],
+        enabled: bool,
+        monotonic: bool,
+    ):
+        self.data = 0
+        self.value_type = value_type
+        self.enabled = enabled
+        self.monotonic = monotonic
+
+    def update(self, value: metrics_api.ValueType) -> None:
+        if not self.enabled:
+            return
+        if self.monotonic and value < 0:
+            logger.warning("Monotonic gauge cannot descend.")
+            return
+        if not isinstance(value, self.value_type):
+            logger.warning(
+                "Invalid value passed for %s", self.value_type.__name__
+            )
+            return
+        self.data = value
+
+    def set(self, value: metrics_api.ValueType) -> None:
+        """See `opentelemetry.metrics.GaugeHandle._set`."""
+        self.update(value)
+
+
+class MeasureHandle(metrics_api.MeasureHandle):
+    def __init__(
+        self,
+        value_type: Union[Type[float], Type[int]],
+        enabled: bool,
+        monotonic: bool,
+    ):
+        self.data = 0
+        self.value_type = value_type
+        self.enabled = enabled
+        self.monotonic = monotonic
+
+    def update(self, value: metrics_api.ValueType) -> None:
+        if not self.enabled:
```

Comment:
Probbably all these checks could be moved in some common (global or base class) function.

Reply:
I created a base class `BaseHandle` in the sdk that has this function.